sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/CosmosItemSerializerTest.java

```diff
@@ -0,0 +1,771 @@
+    public CosmosItemSerializerTest(
+        CosmosClientBuilder clientBuilder,
+        boolean inContentOnWriteEnabled,
+        boolean nonIdempotentWriteRetriesEnabled,
+        boolean useTrackingIdForCreateAndReplace) {
+        super(clientBuilder);
+
+        this.isContentOnWriteEnabled = inContentOnWriteEnabled;
+        this.nonIdempotentWriteRetriesEnabled = nonIdempotentWriteRetriesEnabled;
+        this.useTrackingIdForCreateAndReplace = useTrackingIdForCreateAndReplace;
+    }
+
+    @DataProvider
+    public static Object[][] clientBuildersWithDirectSessionIncludeComputeGatewayAndDifferentItemSerializers() {
+        boolean[] contentResponseOnWriteValues = new boolean[] { true, false };
+        boolean[] nonIdempotentWriteRetriesEnabledValues = new boolean[] { true, false };
+        boolean[] trackingIdUsageForWriteRetriesEnabledValues = new boolean[] { true, false };
+
+        CosmosItemSerializer[] itemSerializers = new CosmosItemSerializer[] {
+            null,
+            CosmosItemSerializer.DEFAULT_SERIALIZER,
+            EnvelopWrappingItemSerializer.INSTANCE_NO_TRACKING_ID_VALIDATION
+        };
+
+        List<Object[]> providers = new ArrayList<>();
+        for (CosmosItemSerializer serializer : itemSerializers) {
+            for (boolean isContentResponseOnWriteEnabled : contentResponseOnWriteValues) {
+                for (boolean nonIdempotentWriteRetriesEnabled : nonIdempotentWriteRetriesEnabledValues) {
+                    for (boolean trackingIdUsageForWriteRetriesEnabled : trackingIdUsageForWriteRetriesEnabledValues) {
+                        if (!nonIdempotentWriteRetriesEnabled && trackingIdUsageForWriteRetriesEnabled) {
+                            continue;
+                        }
+
+                        Object[][] originalProviders = clientBuildersWithDirectSession(
+                            isContentResponseOnWriteEnabled,
+                            true,
+                            toArray(protocols));
+                        List<Object[]> providersCurrentTestCase = new ArrayList<>();
+
+                        for (Object[] current : originalProviders) {
+                            Object[] injectedProviderParameters = new Object[4];
+                            injectedProviderParameters[0] = current[0];
+
+                            injectedProviderParameters[1] = isContentResponseOnWriteEnabled;
+                            injectedProviderParameters[2] = nonIdempotentWriteRetriesEnabled;
+                            injectedProviderParameters[3] = trackingIdUsageForWriteRetriesEnabled;
+                            providersCurrentTestCase.add(injectedProviderParameters);
+                        }
+
+                        CosmosClientBuilder builder = createGatewayRxDocumentClient(
+                            TestConfigurations.HOST.replace(ROUTING_GATEWAY_EMULATOR_PORT, COMPUTE_GATEWAY_EMULATOR_PORT),
```

Comment:
What is the rationale behind the port switch - is it because more and more accounts are deployed on Compute Gateway?

Reply:
So, you have 3 flavors of CosmosClientBuilder - direct, one to route requests to the compute gateway endpoint and routing gateway endpoint?

Reply:
Yes, correct